### PR TITLE
[SSO] wrap create user in transaction.atomic()

### PR DESCRIPTION
**Pull request by biyeun**
## Summary
One of our pilot testers triggered a very weird set of errors ([seen here](https://sentry.io/organizations/dimagi/issues/2473792575/?environment=staging&project=136860&query=is%3Aunresolved+sso&statsPeriod=14d)). It's strange because they could have only gotten to the create user block if a `User` for them did not exist. However, when the process to create a new user is triggered, an `IntegrityError` is thrown indicating that the `WebUser` already exists, yet a few lines later there is another `DoesNotExist` when trying to fetch the `User` with the given username. This has worked flawlessly in local testing and with QA, so some strange race condition is happening.

Anyway, generally when I see `IntegrityErrors` that defy explanation the fix is using a `transaction.atomic()` block. If anyone else has other ideas, my ears are open.

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage
SSO is well tested, especially in this area. However, I'm not sure how to replicate this race condition in tests.

### QA Plan
QA is actively happening on staging, but this can be merged in as soon as it's approved because it touches an area of the codebase which is not active on production.

### Safety story
This area of the codebase is not active on production.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### CODE REVIEW

**Review by millerdev — COMMENTED** [review]
Is it possible to test the scenario you are hypothesizing about? By that I mean, possible to reproduce the race condition. In the past I have used monkey-patching + threading + locks to ensure that code is executed in precisely the order needed to reproduce a race. Maybe that would be useful here?

File: corehq/apps/sso/backends.py [hunk 1]
```diff
@@ -76,8 +77,26 @@ def authenticate(self, request, username, idp_slug, is_handshake_successful):
...
             web_user = WebUser.get_by_username(username)
         except User.DoesNotExist:
-            user, web_user = self._create_new_user(request, username, async_signup)
-            is_new_user = True
+            try:
+                # the atomic block is necessary to avoid any race conditions
+                # that came up during testing of this feature.
+                with transaction.atomic():
+                    user, web_user = self._create_new_user(request, username, async_signup)
+                    is_new_user = True
```
Comment (mjriley) [inline corehq/apps/sso/backends.py:85]: If `self._create_new_user()` did need an atomic block, it should be inside the method, not enclosing it. The atomic block is meant to ensure multiple changes all happen together, and here, the only thing that is happening is that a new user is created. It's possible that the operations in `_create_new_user()` need to all happen together, but then that means I'd never want to call `_create_new_user()` _without_ an atomic block, due to the possibility of another race condition. We eliminate that possibility for all callers by moving the atomic block to where the multiple operations happen.

File: corehq/apps/sso/backends.py [hunk 2]
```diff
@@ -76,8 +77,26 @@ def authenticate(self, request, username, idp_slug, is_handshake_successful):
             is_new_user = False
             web_user = WebUser.get_by_username(username)
         except User.DoesNotExist:
-            user, web_user = self._create_new_user(request, username, async_signup)
-            is_new_user = True
+            try:
+                # the atomic block is necessary to avoid any race conditions
+                # that came up during testing of this feature.
```
Comment (mjriley) [inline corehq/apps/sso/backends.py:82]: I would prefer less confidence here. When I come upon a comment like this, and it doesn't match my assumptions about how the code works, I defer to the comment. My concern here is that this hasn't been proven as a solution, and could lead a future developer in the wrong direction if this theory is incorrect.

**Review by mjriley — CHANGES_REQUESTED** [review]
I might be reading the code incorrectly, but I don't think we've fixed the problem here. From the sentry entry, after creating this new transaction block, we'd be hitting an integrity error on the very first attempt at creating a user.

Looking above the transaction, we query the `User` table. Because we're in this block of code, presumably no user was found. However, it's a network environment -- during the time we've queried and found no user, and when we actually attempt to create that user, a separate request could have come in and generated this user, hence the integrity constraint. To prevent this situation in the future, enclose both the check for the existing user and the subsequent creation of the new user in an atomic block -- that way they can never be out of sync.

Does that sound like something our tests could have done? Would spamming the submit button have been able to send back-to-back requests that got processed out of order, or perhaps this gets triggered in the background?

**Pull request closed without merging**